docs/proposals/improved-schema-for-services.md

```diff
@@ -0,0 +1,472 @@
+Getting that metadata from the cluster is challenging because a normal cluster user (which is a normal user in vanilla k8s or non-admin admin user for openshift) doesn't have access to that metadata so we are gonna follow a 3 tiered approach
+
+- check if the user has access to the CRD, if so then use that. The reason we are doing this is because a non-admin user can still have access to CRDs because they might be given explict permissions by the admin.
+- get the `swagger.json` from the openapi endpoint provided by kubernetes. a sample can be found here https://gist.github.com/girishramnani/f29949d5cb8c6547102776437e05ac19
+- finally if we cannot get the openapi data as well then use the `ClusterServiceVersion` to get the parameters.
+
+This change would affect multiple service commands and the changes are described briefly below -
+- `odo catalog describe service` should include the metadata fields with description so the users can provide then when doing `odo service create`. 
+- `odo catalog list service` wouldn't change at all.
+- `odo service create` should take flags and dynamically fill the CRD structs - metadata would be used for validation
+- `odo service create --from-file` would be used by adapter team or whoever wants to provide the whole CRD themselves as a file. this feature is already present and working.
+- `json` output for all commands.
+
+# Implementation plan
+
+Note - Below I would mention `cluster` manytimes and that means both openshift and kuberenetes. If there is something specific to one or the other then I would mention it.
+The actually order in which these features would be implemented in odo would as follows -
+
+- [ ] `ClusterServiceVersion` approach 
+- [ ] using `swagger.json` to get `openAPIv3schema`
+- [ ] admin-only approach of using `CRD` api from the cluster.
+
+
+Below explains the flow, that is how the code would get executed - 
+
+## Getting the metadata
+
+### Admin Access/Access to CRD API path
+First we would consider the scenerio where the user has admin access or privileges to access the CRD API to the cluster.
+Cluster provides an api to get metadata for any CRD needed. URL - `<cluster-url>/api/kubernetes/apis/apiextensions.k8s.io/v1/customresourcedefinitions`. 
+
+Sample output - 
+https://gist.github.com/girishramnani/cbb4400e463efe89c13f1386e0788793
+
+We care about the `openAPIV3Schema` as that would be used to build the CRD struct. 
+
+### Kuberenetes cluster Swagger has the schema
+
+If the user doesn't have CRD access or that portion is not implemented yet then we fetch try to fetch the same `openAPIV3Schema` from the cluster's `swagger.json` from the endpoint `<cluster-url>/api/kubernetes/openapi/v2`. This is a very large document as it holds all the definitions present on the cluster.
+
+So this needs to be cached and refreshed whenever a new operator is installed.
```

Comment:
I'm not very comfortable with managing an automatic cache (which, without a serious protocol between the two parts, is doomed to be inefficient).

What about a specific and explicit odo command, that the user could run to fetch the param information for one or more services?

User story:

1. `odo catalog list services` indicates the params or number of params if known, or if not known, that the user can fetch them.
2. fetch the params for a specific service 

```
$ odo catalog list services
Services available through Operators
NAME                                             CRDs                                                          Params
etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore       (to fetch)
service-binding-operator.v0.7.1     ServiceBinding                                             3 parameters
To fetch params for a service, you can run `odo service fetch-params <service-name>


$ odo service fetch-params etcdoperator.v0.9.4-clusterwide/EtcdCluster
...
```


Also, only the params information could be stored locally (and not the complete swagger/CRD info), independently of the source of the params (ClusterServiceVersion, swagger, CRD), so the use of this data by `odo service create` is generic.

The user could detect which method for fetching the data is available, or give the user the choice to select one.

Reply:
> fetch the params for a specific service

I'm not sure about this. As a user I would be pretty annoyed that I have to run extra command to get information that is essentially required for creating services.
Why `odo catalog list services` couldn't do that automatically?



> Also, only the params information could be stored locally (and not the complete swagger/CRD info), independently of the source of the params (ClusterServiceVersion, swagger, CRD), so the use of this data by `odo service create` is generic.

I like the idea of caching just params for all sources. This is good.

> The user could detect which method for fetching the data is available, or give the user the choice to select one.

I think that this is exposing implementation to users. The user doesn't care where the data comes from. Especially when it can be a fairly complicated decision.